pilz_control/include/pilz_control/pilz_joint_trajectory_controller_impl.h

```diff
@@ -178,6 +180,39 @@ triggerMovementToHoldPosition()
   JointTrajectoryController::setHoldPosition(this->time_data_.readFromRT()->uptime);
 }
 
+template <class SegmentImpl, class HardwareInterface>
+inline bool PilzJointTrajectoryController<SegmentImpl, HardwareInterface>::
+checkStates(const std::vector<double>& old_desired_position,
+            const std::vector<double>& new_desired_positioin,
+            const ros::Duration& period) const
+{
+  return (active_mode_ != Mode::HOLD
+          && !cartesian_speed_monitor_->cartesianSpeedIsBelowLimit(old_desired_position,
+                                                                   new_desired_positioin,
+                                                                   period.toSec(),
+                                                                   0.25 /*limit */));  // TODO: HSL
+}
+
+template <class SegmentImpl, class HardwareInterface>
+void PilzJointTrajectoryController<SegmentImpl, HardwareInterface>::
+reactToFailedStateCheck(const ros::Time& updated_uptime, const Trajectory& curr_traj)
+{
+  switchToHoldMode();
+  JointTrajectoryController::preemptActiveGoal();
+  // TODO: Update desired state
+  triggerMovementToHoldPosition();
+}
+
+template <class SegmentImpl, class HardwareInterface>
+inline void PilzJointTrajectoryController<SegmentImpl, HardwareInterface>::
+switchToHoldMode()
```

Comment:
Why do you omit the lock used here:
> handleHoldRequest(std_srvs::TriggerRequest&, std_srvs::TriggerResponse& response)
{
  std::lock_guard\<std::mutex> lock(sync_mutex_);

 > if(active_mode_ == Mode::HOLD)

Reply:
First off: Thanks for pointing this out.
You actually made me aware of the fact that we are facing a much more general problem:
How should we deal with mutexes in the `update()` function?
Let's for example have look at [`sync_mutex_`](https://github.com/PilzDE/pilz_robots/blob/melodic-devel/pilz_control/include/pilz_control/pilz_joint_trajectory_controller.h#L129) in the `PilzJointTrajectoryController` class. The `sync_mutex_` mutex can be locked as long as it takes to perform a complete stop (see [`handleHoldRequest` function](https://github.com/PilzDE/pilz_robots/blob/melodic-devel/pilz_control/include/pilz_control/pilz_joint_trajectory_controller_impl.h#L111)). This can be a multiple of the actual controller period which, therefore, violates the "realtime" aspect if the mutex is used to change the value of the `active_mode_` member.
But even if I for the moment neglect the `sync_mutex_` mutex, let's have a look at the member [`curr_trajectory_box_`](https://github.com/ros-controls/ros_controllers/blob/melodic-devel/joint_trajectory_controller/include/joint_trajectory_controller/joint_trajectory_controller.h#L196) of the `JointTrajectoryController` base class. The `curr_trajectory_box_` member is e.g. needed to [set the stop trajectory](https://github.com/ros-controls/ros_controllers/blob/melodic-devel/joint_trajectory_controller/include/joint_trajectory_controller/joint_trajectory_controller_impl.h#L811) in case of a failed state check, internally also uses a mutex. Who does guarantee that the member is not locked for to long. Is this guaranteed via code review???